hazelcast/include/hazelcast/client/connection/ClientConnectionManagerImpl.h

```diff
@@ -0,0 +1,354 @@
+
+                std::auto_ptr<protocol::ClientMessage>
+                encodeAuthenticationRequest(bool asOwner, serialization::pimpl::SerializationService &ss,
+                                            const protocol::Principal *principal);
+
+                void onAuthenticated(const Address &target, const boost::shared_ptr<Connection> &connection);
+
+                void fireConnectionAddedEvent(const boost::shared_ptr<Connection> &connection);
+
+                void removeFromActiveConnections(const boost::shared_ptr<Connection> &connection);
+
+                void fireConnectionRemovedEvent(const boost::shared_ptr<Connection> &connection);
+
+                void disconnectFromCluster(const boost::shared_ptr<Connection> &connection);
+
+                boost::shared_ptr<util::impl::SimpleExecutorService>
+                createSingleThreadExecutorService(spi::ClientContext &client);
+
+                void fireConnectionEvent(const hazelcast::client::LifecycleEvent::LifeCycleState &state);
+
+                void connectToClusterInternal();
+
+                std::set<Address> getPossibleMemberAddresses();
+
+                std::auto_ptr<ClientConnectionStrategy> initializeStrategy(spi::ClientContext &client);
+
+                void startEventLoopGroup();
+
+                void stopEventLoopGroup();
+
+                class InitConnectionTask : public util::Runnable {
+                public:
+                    InitConnectionTask(const Address &target, const bool asOwner,
+                                       const boost::shared_ptr<AuthenticationFuture> &future,
+                                       ClientConnectionManagerImpl &connectionManager);
+
+                    void run();
+
+                    const std::string getName() const;
+
+                private:
+                    boost::shared_ptr<Connection> getConnection(const Address &target);
+
+                    const Address target;
+                    const bool asOwner;
+                    boost::shared_ptr<AuthenticationFuture> future;
+                    ClientConnectionManagerImpl &connectionManager;
+                };
+
+                class AuthCallback : public impl::ExecutionCallback<boost::shared_ptr<protocol::ClientMessage> > {
+                public:
+                    AuthCallback(const boost::shared_ptr<Connection> &connection, bool asOwner, const Address &target,
+                                 boost::shared_ptr<AuthenticationFuture> &future,
+                                 ClientConnectionManagerImpl &connectionManager);
+
+                    virtual void onResponse(const boost::shared_ptr<protocol::ClientMessage> &response);
+
+                    virtual void onFailure(const boost::shared_ptr<exception::IException> &e);
+
+                private:
+                    const boost::shared_ptr<Connection> connection;
+                    bool asOwner;
```

Comment:
This value should be atomic. I have checked the java side. It seems it is wrong there too. I will make it volatile in java.

Reply:
Changed to AtomicBoolean.

Reply:
This is supposed to be the field in Connection class. Change to AtomicBoolean in Connection.